### ISSUE TITLE
First attempt at implementing a generalized extraction configuration schema and parsing logic. [phab:D2243]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2707]
_Created by **craigbarber** at 2018-04-05 00:22:21_: 

(This is an incremental CL)
The goal here was to implement an intermediate configuration schema which can be used to generate a customized extraction container tailored for the individual repo. This schema would be the output of both the automated repo examination logic as well as a parser for a separate user friendly schema which could live in the repo itself. In both cases, the common patterns you see in the mvn_config.json file could be stored in a mapping used to generate this intermediate form.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-05 00:26:38_:

  - incorporated linter feedback

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

Hmm maybe is there some way to indicate what the format of the output should be?  (Is there some file or webpage or something that dictates the format?)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

Is this intended to ever run in a directory where we don't have complete control of the contents?  (Could the name collide?)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

My go is still weak - don't suppose there's a joiner that does this?

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

ah maybe a slightly more descriptive name (or at least not such a generic name) - I had already forgotten what the regex did 6 lines later when reading this

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

why is this (and the ones above) log.Fatalf instead of t.Fatalf?

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

I don't know if we follow style guide that says you must have newline at end of file or not

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

Is it confusing to have both of these files named config_parser.go?  I dunno.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 16:48:32_:

Should we be reusing analysis.proto or make a separate proto?

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-05 17:03:22_:

  - incorporated linter feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

My goal is to make the factor images as orthogonal as possible, however in some situations some cross deps become necessary. Having a mvn/java cross dep seems reasonable in this case IMHO.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

The goal here is to have this schema be the format outputted by automation which examines the repo teasing out extraction requirements.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

My thoughts here are to have a separate mapping containing some of these common config patterns. This mapping can be utilized by the automation which examines repo contents as well as the parser which consumes the more user friendly manual config. Both these systems are to follow.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

In this case this is for a container image we have complete control over. The volumes themselves are basically mount points for I/O when running the container. The names of the volumes are arbitrary and don't correspond with real directories outside of the container. See: https://docs.docker.com/storage/volumes/

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

There is a join, however not one that also provides for the sandwiching of the values within the parens above, that I know of.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

My understanding of the go convention, (which could be wrong), is log.Fatal is for unexpected errors, whereas t.Fatal is meant for actual test failures.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

Added for good measure.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

Yeah I'm not sure what the convention is in Go for naming executables and their corresponding binaries. Will have to defer to our local golang guru for some guidance on that.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-05 17:04:23_:

Good question, I'm not sure. Would appreciate guidance from others on this as well.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Blank above, please.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Each package should have a package comment, e.g.,

    // Package config defines configuration settings for extracting compilation records from
    // a repository using the Maven build system.
    … etc. as needed …

This becomes the package description in http://godoc.org/, so it's helpful to have this be explanatory.

N.B.: Only one of the files in the package needs to have such a comment.

https://github.com/golang/go/wiki/CodeReviewComments#package-comments

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Protobuf message types should be passed by their pointer. The interesting methods on the generated code are attached to the pointer, not the bare type: `config *apb.ExtractionConfigSchema`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Group related values together:

```
const (
    // DefaultBaseImage is the name of the default base Docker image to use for extraction.
   DefaultBaseImage = "debian:jessie"

   ...
)
```

Please also make comments into complete sentences, per https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences, rather than JavaDoc-style "name Telegraphic description".

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

I think you could leave off the "Var" suffix here.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Here and throughout: Variable names living in small scopes need not be quite so verbosely-named. I've made some specific suggestions below, but as a rule of thumb try to avoid echoing type information in the variable names.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

A typical naming convention in Go for a constructor-type function is to start with `New…`. Here, I'd suggest `NewImage` (or `NewExtractionImage` if you expect there to be other kinds of images in the config as well).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`factorImage` → `image`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

This is fine as written, but could probably be a single top-level string, since all the substituents are constant:

```
var baseConfig = fmt.Sprintf(`FROM %[1]s
VOLUME %[2]s
ENV %[3]s=%[2]s
VOLUME %[4]s
ENV %[5]s=%[4]s`, DefaultBaseImage, DefaultRepoVolume, RepoVolumeEnv, DefaultOutputVolume, OutputVolumeEnv)

…
fmt.Fprintln(&buf, baseConfig)
```

Also: If these config values aren't referenced outside the package, they could be unexported (e.g., `defaultBaseImage`, etc.)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`imageCopyArtifact` → `artifact`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

`fmt.Errorf("missing source for copy artifact: %s", image.ImageUri)`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`imageEnvVar` → `env`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`runCmd` → `cmd`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

I'm not sure why this comment is here, since this is only extracting some bytes from the buffer—was this meant to be part of the function doc?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

I think we should create a separate proto file for these, rather than putting them in analysis.proto.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Why "Schema" vs., say, "ExtractionConfiguration"?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

I recommend nesting most of these subordinate messages inside the top-level config message, unless they are going to be used separately.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

I suggest `destination` or `target` (but not to abbreviate the field name in this case; it's not that long).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

The names of these messages should be simplified: If they are at the top-level, they should be generic enough not to reference FactorImage, and if they're specific to a FactorImage they can be nested inside it to avoid the need for repeating the prefix:

Suggestions

* FactorImageEnvVar → EnvVar, or even just Env (which is what we used in CompilationUnit, for example)
* FactorImageCopyArtifact → CopyArtifact, or even just Copy (since Artifact doesn't really add any information; maybe "Spec" would be better?)
* FactorImage → Image (what does "Factor" mean here; these are just ordinary Docker images, no?)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Let's just use the proto package name here: `kythe.proto.ExtractionConfiguration` or whatever you wind up calling it.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`extractionConfig` → `config`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

```
if err := ioutil.WriteFile(…); err != nil {
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Unexport?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

This function doesn't load anything.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

`filepath.Join(testDataDir, fileName)`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Does this actually work in `bazel test`? Ordinarily we need to plumb through `TEST_SRCDIR` to find the root of the test execution path. The Blaze docs should say what the assumptions are.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

In a test, it's OK to have setup functions crash if they don't work, which simplifies the test logic:

```
func mustLoadConfig(path string) *apb.ExtractionConfiguration {
    data, err := ioutil.ReadFile(filepath.Join(testDir, path))
    if err != nil {
       log.Panicf("Unable to read test file: %v", err)
    }
    …
    return config
}
```

This isn't kosher for a package API, but is fine in a test where a test setup is a failure you want to catch anyway.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:32:26_:

Don't echo types in the name, please.

This can also be boosted to the top level:

    var multipleNewlines = regexp.MustCompile(`\n{2,}`)

then

    multipleNewlines.ReplaceAllString(image, "\n")

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-05 17:45:19_:

Oh, another related question: If these are command-lines, should we separate the arguments? If not, maybe include a note in the doc about what the escaping rules are (e.g., "these command lines are passed literally to the docker RUN atom, so values must be escaped appropriately inline")

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-05 17:46:00_:

Yea I mean you could make a new type or such thing where its toString surrounds itself with " ", and then join on thing mapped to those, but maybe that's overkill compared to just doing this.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-06 21:27:04_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-06 21:30:24_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-06 21:30:41_:

The test currently works as is within bazel, however if this will be an issue with blaze and google3, I'm all for taking pro-active measures. Added $TEST_SRCDIR as a path prefix to testDataDir. PTAL

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-06 21:30:41_:

Added logic to escape the args.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-06 21:30:41_:

Changed all but FactorImage. "Factor" in this case indicates that these images are going to be used as factors which are combined to form a composite image as the output of the this library.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-06 21:30:41_:

Agreed that is an approach I could take, but yeah it feels like overkill in this case.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-06 22:09:55_:

(nit on no newline)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-06 22:09:55_:

Is it worth linking to some doc on how docker images are set up, or would that not make sense here?  I think it might be worthy to ground some of the terminology (factor image, run, their uri spec, etc)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-06 22:09:55_:

I see this was marked as done, but is it not done?

imageString
expectedImageString

both still using Hungarian notation

I don't remember the original anymore (yay phabricator), so maybe this is fine now?

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-06 22:29:55_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-06 22:30:12_:

Good catch, I had noticed the renaming of the regexp but not the string vars.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-09 18:22:28_:

Pinging this. Are there any outstanding issues for this?

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-09 21:10:25_:

Fine by me, dunno if anyone else still has open concerns.

**Comment by kythe** [comment]
_**danielmoy-google** accepted this revision at 2018-04-09 21:10:25_

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-09 21:11:08_:

(I'm also kinda assuming that since this is still WIP, we can address other concerns in near-future subsequent changes, since it'll be awhile before this goes all the way to production)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

This can be boosted to the package level (so it need only happen once).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

Actually, do these constants need to be exported from the package at all? If not: `defaultBaseImage`, `defaultRepoVolume`, etc.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

`extractionConfig` → `config`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

What does "Factor" mean in this context? Is there some reason this can't simply be "Image"?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

Dead?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-09 21:27:29_:

Within a test use the `t.*` methods. It's OK for a test to have some helper functions that `log.Fatal`, but generally the stuff in scope of a `t` should use it.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-10 00:34:59_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-10 00:36:25_:

If we're able to realize the plan of having either automation logic and/or "user friendly" config logic produce these configs, I think it'll be useful to have these constants exported.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-10 00:38:53_:

Factor in the composite->factor sense. It's meant to denote that each of these images are meant to be a factor composing the generated custom composite image.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-10 00:39:52_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

It wouldn't be hard to export them later, if that happens, but YMMV.

A useful trick for that, if you wish:

    gofmt -w -r 'DefaultBaseImage->defaultBaseImage' kythe/go/extractors/config/config_parser.go

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

I think if you want the variable expanded you'll need to pass it through http://godoc.org/os#ExpandEnv before calling Open on it.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

Not quite done—note `:=` here (it happens to work without because there's already an `err` in the enclosing scope, but the point was not to reuse it)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

Nothing requires the names to match up, but generally I like to follow the convention that `package foo` resides in `foo.go` (assuming it's a single-file package) and that `package main` resides in `binary_name.go` (corresponding to whatever the `name` of the Bazel `go_binary` rule is called).

It's fine to have them overlap, though since library packages are usually //not// named with underscores, that convention avoids the issue.

None of this is graven in stone, though.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

If you indent the example a couple extra spaces inside the comment, godoc will format it verbatim.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

A plea for flowing:

```
// Binary config_parser provides a tool to parse a configuration file the the
// schema defined in kythe.proto.ExtractionConfiguration. Its output is a
// Dockerfile defining an image which can be used to perform an extraction on a
// repository according to the given configuration.
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

`if err := jsonpb.Unmarshal(configFile, &config); err != nil {`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

I see. Do we really need that qualification? I think it's not really adding any information—we have a configuration that refers to some docker images, and their intended use doesn't really seem to matter. (In particular, we don't have `FactorImage` vs. `NonFactorImage` for some other purpose—and even if we did, would they be shaped differently?)

On that basis, what do you think about

    FactorImage → Image
    required_factor_image → required_image

?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-10 02:03:17_:

Please add a newline.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-11 20:21:40_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-11 20:22:45_:

Feedback incorporated. PTAL

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-11 20:56:37_:

^ I think this comment was getting at, we'd rather imports be grouped as:

"bytes"
"fmt"

"kythe.io/some/package/we/own"
"kythe.io/another/package/we/own"

"bitbucket.org/something/we/don't/own"
"google.com/even/stuff/from/el/goog"

apb "kythe.io/some/proto/at/bottom"
opb "github.com/but/also/with/non-kythe/pb"


So you want to reorder:

"bytes"

"bitbucket.org/stuff"

ecb "kythe.io/kythe/stuff"

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-11 20:57:30_:

\n

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:08_:

`log.Fatalf` (because you have %v here)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:08_:

`extractionImageContent` → `image`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:09_:

Yeah, that. It's changed a few times since I made my original comment, so there's more to do now. :)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:09_:

t.Fatalf

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:09_:

t.Fatalf

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:09_:

Avoid "assert" style in Go tests (see https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures for some context).

It's fine to have a function that reports an error that the test can log, but avoid passing the `*testing.T` value into helper functions that actually report test failures. (Logging stuff is fine)

Better: `imagesEqual(got, want []byte) (bool, string) {`

```
if eq, diff := imagesEqual(got, test.want); !eq {
    t.Errorf("Bootstrap images not equal: diff: %s", diff)
}
```

or words to that effect.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-11 21:35:09_:

This is entirely an optional nit that you should feel free to disregard.

Most Unix binaries are verb-like: `ls`, `cat`, `pwd`, `sort`, etc. It's nice for reading command lines if ours do that too, instead of naming them like objects. So how about `parse_binary_config` instead of `binary_config_parser`. After all, what the user cares about is that the tool does so, rather than being so. :)

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-11 23:20:31_:

Understood. I added the "binary" to the name as I interpreted a previous comment to mean that using the suffix "binary_<name>" was a golang convention in regards to a corresponding library <name>. If that's not the case, then I'd rather ditch the "binary" part entirely as that is not in fact what this program is doing, rather it's parsing a JSON config :)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-12 00:41:56_:

My apologies, I was probably not clear enough in my comment. It's not necessary to include `binary_*` in the name at all.

There is a convention that the package comment in `package main` should mention the name of the binary, e.g.,

    // Binary parse_config converts a text-format proto config for an extractor into
    // wire format in the specified output file.

But that's not meant to constrain what you call the binary itself, it's only meant to ensure the comment winds up being a complete sentence.

So yeah: `parse_config` or even just `parseconfig` would be a fine name for this tool if you like that.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-12 17:43:44_:

  - incorporated CL feedback

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-04-12 17:44:04_:

That definitely helps to clear things up. Thanks :)

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-12 17:44:04_:

Incorporated latest feedback. PTAL

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-04-12 17:47:18_:

Thanks, you've been very patient with all these rounds of comments and I'm very grateful for that.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2018-04-12 17:47:18_

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-04-12 18:20:48_:

As scottg is fond of saying, good feedback is a gift. I'm grateful for you all lending me your time and energy on this :)

**Comment by kythe** [comment]
_**craigbarber** closed via commit c5eadb6ed307db9725d1d48c8435e1dde38efd97 at 2018-04-12 21:59:17_
